### ISSUE TITLE
ESTree compatible AST

### ISSUE THREAD

**Issue by sosukesuzuki** [swc-project/swc#2123]
**Describe the feature**

Support [ESTree](https://github.com/estree/estree) compatible AST.

Some options:

- SWC output ESTree compatible AST directly.
- New tool like `swc-to-estree`

**Babel plugin or link to the feature description**

Nothing.

**Additional context**

I'm a maintainer of Prettier Code Formatter and I'm developing to use SWC as a parser for Prettier (Disclaimer: It is a 100% personal activity and not affiliated with any organization.) Prettier already has a code base to handle ESTree compatible ASTs well. So, if SWC supports ESTree, it will be easier. Also, ESTree is the de facto standard for JavaScript ASTs, and by supporting it, many other tools may be able to use SWC.

related: https://github.com/swc-project/swc/issues/1392

**Comment by sosukesuzuki** [comment]
@kdy1 What do you think about this?

**Comment by jdalton** [comment]
@sosukesuzuki Just a heads up, the `parse` APIs [are being deprecated](https://github.com/swc-project/swc/pull/1893#issuecomment-903359692). But [I've asked](https://github.com/swc-project/swc/pull/1893#issuecomment-903359692) if there'll be a way to support this scenario in the future as I've seen some perf improvements on my end _(replacing `acorn` with `swc`+ little AST adapters)_.

**Comment by kdy1** [comment]
I made a mistake. `parse` will be preserved to help some tools like ast explorer.

**Comment by sosukesuzuki** [comment]
Is it expected that third party libraries will only use the swc parser?

**Comment by kdy1** [comment]
No, but rust <-> js is very slow and it overwhelms time used for actual work.

**Comment by sosukesuzuki** [comment]
I see. Out of curiosity, you are doing integration SWC and Next.js at vercel? At there, how to integrate Rust land and JS land?

**Comment by kdy1** [comment]
Yes. We are working on it. We are going to reimplement all transforms in rust, and call `transform()` from js.

**Comment by sosukesuzuki** [comment]
I found https://github.com/vercel/next.js/tree/canary/packages/next/build/swc. It looks good, thanks

**Comment by sosukesuzuki** [comment]
@kdy1 What do you think about making swc's AST consistent with ESTree, which is the de facto standard?

**Comment by kdy1** [comment]
What's the advantage if I make it compatible with estree? rust does not have GC, so estree-like ast will degrade performance.

**Comment by sosukesuzuki** [comment]
I did not know about the performance drawbacks. The advantage is that it is much easier for existing tools such as ESLint and Prettier to use swc as a parser.

**Comment by kdy1** [comment]
I see. But actually, passing ast to javascript from rust is very slow. Parser itself is fast, but json serialization/deserialization is the problematic part, and I'm not sure about the way to make it fast.
Maybe lazy accessors can be a solution, but I didn't try it.

**Comment by jdalton** [comment]
If you provide the string the user can do things like [fast-json](https://github.com/alemures/fast-json#readme). This allows for them to choose which serializer (`JSON.parse` or otherwise) to use.

**Comment by kdy1** [comment]
@jdalton If it's fast enough, I think we can support babel plugins. So I think it is worth benchmarking.

**Comment by jdalton** [comment]
As a test are raw parse power I tried parsing [babylon.max.js](https://cdn.babylonjs.com/babylon.max.js) which s 10.6MB of JS. It takes `acorn` 706ms and `swc` ~2 seconds so I can see what you mean about the Rust->JS transition.

**Comment by overlookmotel** [comment]
Please see #2175 if you're interested. I've done some experiments into speeding up `parse()`.

**Comment by axetroy** [comment]
Why not just write a package to transform SWC's AST to ESTree's AST

```js
const swcAst = parseSync(code);
const estreeAst = swc2estree(swcAst);
```

**Comment by axetroy** [comment]
> Why not just write a package to transform SWC's AST to ESTree's AST
> 
> const swcAst = parseSync(code);
> const estreeAst = swc2estree(swcAst);

There's already have. https://github.com/swc-project/swc/blob/main/crates/swc_estree_compat/README.md

Bug it did not available on npm

**Comment by alexander-akait** [comment]
@kdy1 Just interesting, is `swc_estree_compat` ready to use?

**Comment by kdy1** [comment]
Yes, but need some triage because I need to port `rkyv` to typescript/javascript.

**Comment by NaviTheCoderboi** [comment]
any update?

**Comment by 0xdevalias** [comment]
> it would be very slow at the moment because of JSON.parse of large AST is very slow

> Any ideas how we can solve it?

It was [already mentioned](https://github.com/swc-project/swc/issues/2123#issuecomment-907720026) in this issue by @overlookmotel back in 2021; but there was a HUGE amount of work exploring / testing all of this (that I just read through) in the following issue:

- https://github.com/swc-project/swc/issues/2175

And then there seems to be some more recent snippets / updates related to it in this issue:

- https://github.com/napi-rs/napi-rs/issues/1502

With this part sounding particularly interesting / useful (at least by my naive read):

> We've implemented many of the ideas I outlined above in [Oxc](https://oxc.rs/), and they do indeed produce a very large speed-up (up to 20x).
> 
> However, what we've discovered along the way is that the approach is hard to generalize. The biggest gain comes from removing the serialization step entirely, but that requires storing all transferrable data in an arena allocator, which means all types need to have lifetimes. All structs and enums on Rust side also need to be `#[repr(C)]` to guarantee their memory layout, and you need codegen infrastructure.
>
> _Originally posted by @overlookmotel in https://github.com/napi-rs/napi-rs/issues/1502#issuecomment-3323595526_

And while I didn't look deeply into it, this might also be potentially relevant?

> [..] you might want to look at: https://github.com/GoogleChromeLabs/buffer-backed-object 
>
> _Originally posted by @overlookmotel in https://github.com/napi-rs/napi-rs/issues/1502#issuecomment-3323595526_

While I haven't read through it yet, I think this is probably the next most interesting / relevant issue over on the `oxc` side of things, which may be the continuation of what was referred to above:

- https://github.com/oxc-project/oxc/issues/2409

---

I've only spent the last hour or so skimming through those past threads out of interest, so I'm definitely no expert on this area of things; but it seemed a waste not to link to such in-depth explorations.

**Comment by overlookmotel** [comment]
@0xdevalias Thanks for your interest in my ancient experiments!

kdy and I discussed this when we ran into each other at a conference earlier this year. Sadly, it sounds like the full-on zero-cost serialization scheme we have in Oxc is unlikely to be workable in SWC. The requirements (particularly adding lifetimes to all AST types) is too large a breaking change for a project as highly depended-on as SWC. In Oxc we have the luxury of not being so popular! So we can break things when we need to.

But there are certainly more efficient serialization formats than JSON which might net a decent speed gain (e.g. `rkyv`, which kdy mentioned above).
